### ISSUE TITLE
[Discussion] Design our edu hypertrons bot procedure

### ISSUE THREAD

**Issue by frank-zsy** [hypertrons/hypertrons-wonderland#2]
I want to launch a new plugin for hypertrons which is for education purpose.

The main purposes of this plugin are:
- Teach students how git works, especially about distribution and branch feature.
- Teach students how to use git, especially about sync among repos and resolve conflicts.
- Teach students how to use GitHub, especially about issue and pull.

How this can be implemented:
- After student install the edu-robot on his/her repo, the bot should add several issues to the repo.
  - Welcome issue to introduce the robot and config guide of it.
  - Git learning procedure issue to teach git related knowledge.
  - GitHub learning procedure issue to teach GitHub related knowledge.
  - Open source procedure issue to teach open source related knowledge.
- The issues should have metadata to trace the learning stage of each issue.
- When student finish the prior stage, new comment should be made by the robot to lead student to next stage.

We need to design the basic procedures and find out how many features we still lack in Hypertrons and implement them, after that, the plugin should be quite easy to accomplish.

**Comment by will-ww** [comment]
Great ! I can invove other students from ECNU to try to collaberate with this plugin.  A good chance to connect to bot and git knowledge.

**Comment by liwen-tj** [comment]
Here is a basic [design document](https://shimo.im/docs/ghKQTvjWH6xPyj3r/ ). Maybe we should first discuss this scheme before going into detailed scheme. Any advice is very welcome.

**Comment by frank-zsy** [comment]
Sorry for the late reply.

I have several suggestions

- This plugin is not just for schools or universities, so maybe we don't really need a teacher's repo, all open source related knowledges can be learned from a single repo.
- We need to make sure all the knowledges points should be included in the procedure, like conflicts, we need to make sure student can meet conflicts in the procedure if he/she follows the instruction. And PRs can be made from one branch to another in a single repo, it is not a problem.
- We may need to design some basic local tools to make sure we can know the local status of the repo, I do not sort it out but I think it is quite necessary.
- We need to make sure the mechanism is flexible in the future, the basic procedure is not enough for further features. Like we need to add release procedure, wiki maintain, project management into the learning flow.
- We need to think about some fail safe like what should we do if student edit the issue by hand, if this can break the procedure flow, then we need to find a way to fix the status.

**Comment by liwen-tj** [comment]
In point 3, what do you mean by `basic local tools`?  You mean something that is offline and installed  on student's own computer?

**Comment by sunshinemingo** [comment]
For point 3, I have some ideas:
* Maybe we provide a public repo for users to fork.  Local tools are available in our provided repositories.
* Maybe it is possible to develop a local tool similar to git. When a user clones the repo locally, after entering a specific command on the command line, he can enter the git environment (similar to entering ```python``` to enter the python environment on the command line). At this time, we can obtain local status of the repo, and at the same time, the user can execute the git command in this environment. In this way, if the user executes the git command wrongly, we can also get the error information.
* Or we can develop a local tool for users to edit the content of files in the repo locally so that we can know the local status of the repo.
* Or we can develop a local tool to monitor the local state of the repo.

**Comment by frank-zsy** [comment]
@sunshinemingo Yes, that is what I mean, but the hard point is how to develop the tool.

Actually, we don't need to monitor everything in local repo, we only need to add hooks(pre-commit, pre-push) to the local .git config. Then we can check the status if the user what to make some interaction with origin repo.

The most common way to do this is provide a shell script for Linux and a cmd file for Windows, also we can use different programming language to do this, like Python, but this requires Python env in local repo which I think is reasonable, or we can provide different language version like Node.js.

**Comment by sunshinemingo** [comment]
Hooks can be divided into two groups: client-side and server-side. Client-side hooks are triggered by operations such as committing and merging, while server-side hooks run on network operations such as receiving pushed commits.

**However, client-side hooks are not copied when a user clones the repository.**

**Comment by frank-zsy** [comment]
That is right, but we can make sure it will be inited like we can add scripts to setup hooks when user installs dependencies, I am not sure whether Python can do that but I am quite sure Node.js can in its post install script.

**Comment by sunshinemingo** [comment]
So, since we can get the local status of the repo through hooks, maybe it is possible to ensure conflicts by comparing the status of the local repo and remote repo.

If no conflict occurs when a user submits a PR, the remote repo may be able to modify the corresponding files to ensure that a conflict occurs.

**Comment by frank-zsy** [comment]
We can make sure the conflicts occur if users follow the instruction, we should make conflicts by design. Modify file content automatically would be very confusing for users.

**Comment by sunshinemingo** [comment]
This is the first version of flowchart. Any advice is very welcome.
![edu-bot](https://user-images.githubusercontent.com/37278844/76532726-fde73b00-64b1-11ea-8e64-eebddcb227e0.png)
